🧶 🪶 Refactor loss helper

mberr opened the pull request:
This PR aims to get rid of the loss helper methods in the training loop, by moving these to the loss class(es) and using polymorphism instead of the existing if / else logic. This should also improve type checking, and readability.

mberr committed:
Add first draft of interface

mberr committed:
Add note

mberr committed:
Add default implementation

mberr committed:
Add docstring

cthoyt committed:
More consistent abstraction for compute_loss

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -140,6 +140,8 @@
 from torch.nn import functional
 from torch.nn.modules.loss import _Loss
 
+from pykeen.training.utils import apply_label_smoothing
```
will have to move that into the more general `pykeen.utils` to avoid circular importing

mberr replied:
After the changes, it is only used in `losses.py`.

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -178,6 +180,40 @@ def __init__(self, size_average=None, reduce=None, reduction: str = 'mean'):
+        positive_scores: torch.FloatTensor,
+        negative_scores: torch.FloatTensor,
+        label_smoothing: Optional[float] = None,
+        batch_filter: Optional[torch.BoolTensor] = None,
```
hmm looks like we'll want to wait on #447 to finish this one

mberr replied:
This was already present in `master`'s loss helpers.

cthoyt replied:
🦜

mberr committed:
Remove _label_loss_helper from SLCWA training loop

mberr committed:
Add process_slcwa_scores for MarginRankingLoss

cthoyt committed:
Add intermediate loss helper function

mberr committed:
remove unused loss helpers from slcwa training loop

mberr committed:
Add process_lcwa_scores to MarginRankingLoss

mberr committed:
Remove unused loss helpers from LCWATrainingLoop

mberr committed:
Add process_lcwa_scores to NSSALoss

mberr committed:
Remove loss helpers

cthoyt committed:
Simplify passing

mberr committed:
Remove compute_loss, and add collect_regularization_term for all models

cthoyt committed:
Reuse devices

mberr committed:
Merge remote-tracking branch 'origin/refactor-loss-helper' into refactor-loss-helper

# Conflicts:
#	src/pykeen/models/base.py
#	src/pykeen/models/nbase.py
#	src/pykeen/training/lcwa.py
#	src/pykeen/training/slcwa.py

cthoyt committed:
Move utility function

mberr committed:
Post-merge fixes

mberr committed:
Merge remote-tracking branch 'origin/refactor-loss-helper' into refactor-loss-helper

# Conflicts:
#	src/pykeen/training/lcwa.py
#	src/pykeen/training/slcwa.py

mberr committed:
Move apply_label_smoothing to losses

mberr committed:
Remove outdated has_mr_loss / has_nssa_loss

mberr committed:
Add TODO

mberr committed:
Remove resolved TODO

mberr committed:
Add TODO

cthoyt committed:
Update losses.py

mberr committed:
Move label smoothing test case

mberr committed:
Merge remote-tracking branch 'origin/refactor-loss-helper' into refactor-loss-helper

cthoyt committed:
Update losses.py

mberr committed:
Add test for processing SLCWA scores

cthoyt commented on tests/test_losses.py:
```diff
@@ -4,12 +4,13 @@
 
 import unittest
```
There don't appear to be any tests that apply label smoothing on actual instances of a loss class. Need to add that

mberr replied:
9480589  😇

mberr committed:
Add test for processing LCWA scores

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -178,6 +205,54 @@ def __init__(self, size_average=None, reduce=None, reduction: str = 'mean'):
+        :param label_smoothing:
+            An optional label smoothing parameter.
+        :param batch_filter: shape: (batch_size, num_neg_per_pos)
+            An optional filter. If given, ... TODO: describe
```
describe

mberr committed:
Fix process_slcwa_scores

mberr committed:
Merge remote-tracking branch 'origin/refactor-loss-helper' into refactor-loss-helper

cthoyt committed:
Update test_losses.py

mberr committed:
fix tests

mberr committed:
Merge remote-tracking branch 'origin/refactor-loss-helper' into refactor-loss-helper

cthoyt committed:
Add test for when label smoothing is applied by loss instance

It fails because loss instances do not have a model attribute

mberr committed:
remove redundant assignment

mberr committed:
Merge remote-tracking branch 'origin/refactor-loss-helper' into refactor-loss-helper

mberr committed:
Forward num_entities for label smoothing

mberr committed:
Forward num_entities for label smoothing

mberr committed:
Add error for losses not supporting label smoothing

mberr committed:
update tests

cthoyt committed:
Pass flake8 and darglint

cthoyt committed:
Add notes on failing tests

cthoyt commented on tests/test_training.py:
```diff
@@ -294,6 +294,7 @@ def setUp(self):
 
     def test_early_stopper_best_epoch_model_retrieval(self):
         """Test if the best epoch model is returned when using the early stopper."""
+        # TODO this test is failing
```
@mberr this test is failing

cthoyt replied:
since this is a higher level test, it would be good for us to make a smaller unit test with fewer components in it that causes the same incorrect behavior

cthoyt commented on tests/training/test_utils.py:
```diff
@@ -47,82 +47,22 @@ def setUp(self):
 
     def test_lcwa_margin_ranking_loss_helper(self):
         """Test if output is correct for the LCWA training loop use case."""
-        factory = TriplesFactory.from_labeled_triples(triples=self.triples)
```
@mberr this test was failing because of the interface change - I think we could either delete it or just move it as an explicitit test in the losses tests

cthoyt committed:
Merge branch 'master' into refactor-loss-helper

cthoyt committed:
Update test_utils.py

Trigger CI

cthoyt committed:
Update losses.py

Trigger CI

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -178,6 +213,61 @@ def __init__(self, size_average=None, reduce=None, reduction: str = 'mean'):
+            A scalar loss term.
+        """
+        # Stack predictions
+        # TODO: fix this, once the negative batches have appropriate size
```
TODO

mberr replied:
0612594

mberr committed:
Fix NSSALoss process_slcwa_scores

cthoyt committed:
Update losses.py

cthoyt committed:
Merge branch 'master' into refactor-loss-helper

mberr commented on src/pykeen/losses.py:
```diff
@@ -259,11 +259,13 @@ def process_lcwa_scores(
         """Process scores from LCWA training loop."""
         # Apply label smoothing
-        if label_smoothing is not None and num_entities is not None and label_smoothing > 0.:
+        if label_smoothing is not None and label_smoothing > 0.:
```
can we move this part into the `apply_label_smoothing` method? iirc we had similar code in other places, too.

cthoyt replied:
Yeah that would be fine

cthoyt replied:
5636fa9b

cthoyt committed:
Merge branch 'master' into refactor-loss-helper

cthoyt committed:
Update losses.py

cthoyt committed:
Pass darglint

cthoyt committed:
Update exception

cthoyt committed:
Update access pattern for loss

Trigger CI

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -295,6 +393,53 @@ def __init__(
+        positive_scores = positive_scores.unsqueeze(dim=1)
+
+        if batch_filter is not None:
+            # negative_scores have already been filtered in the sampler!
```
So can this logic be removed from the process scores functions? Right now the batch_filter isn't used for anything else

cthoyt replied:
used in NSSA loss, can't avoid

mberr committed:
Improve docstring for process_lcwa_scores

mberr committed:
Add more todos

mberr committed:
Add another todo

mberr committed:
Fix NSSA for filtered negatives

mberr committed:
Add todo

mberr committed:
Update NSSA loss

mberr committed:
add todo

mberr committed:
Add test for filtered slcwa batches

cthoyt commented on tests/cases.py:
```diff
@@ -192,19 +192,39 @@ def _check_loss_value(self, loss_value: torch.FloatTensor) -> None:
+        self,
+        positive_scores: torch.FloatTensor,
+        negative_scores: torch.FloatTensor,
+        batch_filter: torch.BoolTensor = None,
```
```suggestion
        batch_filter: Optional[torch.BoolTensor] = None,
```

mberr committed:
fix nan in NSSA lcwa scores

mberr committed:
Remove outdated tests

mberr committed:
update custom NSSA test

Trigger CI

mberr committed:
Update tests/cases.py

Co-authored-by: Charles Tapley Hoyt <cthoyt@gmail.com>

PyKEEN-bot committed:
Trigger CI

mberr commented on src/pykeen/losses.py:
```diff
@@ -178,6 +224,75 @@ def __init__(self, size_average=None, reduce=None, reduction: str = 'mean'):
+
+        :param positive_scores: shape: (batch_size, 1)
+            The scores for positive triples.
+        :param negative_scores: shape: (batch_size, num_neg_per_pos)  # TODO: Pre-filtered? => shape: (nnz,)
```
open todo

mberr commented on src/pykeen/losses.py:
```diff
@@ -178,6 +224,75 @@ def __init__(self, size_average=None, reduce=None, reduction: str = 'mean'):
+        :param label_smoothing:
+            An optional label smoothing parameter.
+        :param batch_filter: shape: (batch_size, num_neg_per_pos)
+            An optional filter of scores to keep. If given, ... TODO: describe
```
open todo

mberr commented on src/pykeen/losses.py:
```diff
@@ -178,6 +224,75 @@ def __init__(self, size_average=None, reduce=None, reduction: str = 'mean'):
+        """
+        Process scores from LCWA training loop.
+
+        # TODO: Input shapes are different for default delegation from slcwa
```
open todo

mberr commented on src/pykeen/losses.py:
```diff
@@ -384,25 +546,97 @@ def __init__(self, margin: float = 9.0, adversarial_temperature: float = 1.0, re
         .. note:: The default hyperparameters are based the experiments for FB15K-237 in [sun2019]_.
         """
+        # TODO: The "temperature" parameter here is more an reciprocal temperature, following the standard nomenclature
+        #  for softmax temperature
```
@cthoyt I remember we discussed this at some point already.

cthoyt replied:
I think the end of the last discussion was we don't want to change it and worry about comparability of data yet

mberr commented on tests/test_training.py:
```diff
@@ -294,6 +294,7 @@ def setUp(self):
 
     def test_early_stopper_best_epoch_model_retrieval(self):
         """Test if the best epoch model is returned when using the early stopper."""
+        # TODO this test is failing
```
this todo seems to be outdated?

cthoyt committed:
Update test_training.py

cthoyt committed:
Merge branch 'master' into refactor-loss-helper

PyKEEN-bot committed:
Trigger CI

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -384,25 +546,97 @@ def __init__(self, margin: float = 9.0, adversarial_temperature: float = 1.0, re
+        neg_loss = functional.logsigmoid(-neg_scores - self.margin)
+        neg_loss = neg_weights * neg_loss
+        neg_loss = self._reduction_method(neg_loss)
+        pos_loss = functional.logsigmoid(self.margin - -pos_scores)
```
double minus?

cthoyt committed:
Merge branch 'master' into refactor-loss-helper

mberr committed:
Update process_slcwa_scores docstring

mberr commented on src/pykeen/losses.py:
```diff
@@ -427,20 +661,11 @@ def forward(
     if cls.synonyms is not None
     for synonym in cls.synonyms
 }
+# TODO: Can we use Resolver.from_subclasses?
```
@cthoyt any reason why we can't? We could also set the synonym at this place.

cthoyt replied:
we could do this, I guess since we're here we might as well. I'll take care of it before we finish this PR

mberr committed:
Fix typo and extend error message

mberr committed:
fix comment

mberr committed:
Remove delegation from slcwa to lcwa scores

mberr committed:
Replace margin activation lookup by resolver

mberr committed:
Add docstring to MarginRankingLoss forward

mberr committed:
Add some comments to MarginRankingLoss's process_lcwa_scores

mberr committed:
do not explicitly pass the default values to torch.nn.Softplus

These values do only control at which point the implementation switches to a linear function to avoid numerical issues. Thus, we opt for letting the PyTorch devs decide on suitable thresholds, and change them in the future.

mberr committed:
Add formula to MarginRankingLoss class' docstring

mberr committed:
extend docstring of Softplus loss

mberr committed:
inline variable

mberr committed:
Remove unused kwargs from CE.forward

mberr committed:
Move TODO to docstring. Also rename internal variable

mberr committed:
Add comments to NSSA and simplify formula

mberr committed:
Add alternative suggestion for loss resolver

mberr committed:
Add test for optimizing with loss

mberr committed:
Add same optimization direction test for slcwa

mberr committed:
update tests

Trigger CI

mberr committed:
Update README

Trigger CI

cthoyt commented on tests/cases.py:
```diff
@@ -185,6 +193,93 @@ def _check_loss_value(self, loss_value: torch.FloatTensor) -> None:
+        assert (negative_scores < positive_scores.unsqueeze(dim=1) - 1.0e-06).all()
+
 
 # TODO update
```
@mberr we already updated this, just forgot to delete the comment :)

cthoyt committed:
Remove old test

cthoyt committed:
Use from_subclasses

Trigger CI

cthoyt committed:
Update cases.py

Trigger CI

cthoyt committed:
Update losses.py

Trigger CI

cthoyt committed:
Fix docstring

cthoyt committed:
Fix typing

Trigger CI

cthoyt commented on src/pykeen/losses.py:
```diff
@@ -267,12 +396,15 @@ class MarginRankingLoss(PairwiseLoss):
 
     hpo_default: ClassVar[Mapping[str, Any]] = dict(
         margin=dict(type=int, low=0, high=3, q=1),
+        # todo(cthoyt): how to we specify choices? Can we directly create them from the
```
Oh this was actually relevant for #461 where I just added some docs for this! We can come back to updating this part in that PR since it's not core to this one

cthoyt committed:
Update loss resolver construction

cthoyt committed:
Use resolver for choices directly

Trigger CI

cthoyt committed:
Update losses.py

Trigger CI
